File: Algorithms/src/916. Word Subsets/Code/solution.java
class Solution {
    public List<String> wordSubsets(String[] words1, String[] words2) {
        int[] maxFreq = new int[26];

        // Precompute the maximum frequency for each character in words2
        for (String word : words2) {
            int[] freq = new int[26];
            for (char c : word.toCharArray())
                freq[c - 'a']++;
            for (int i = 0; i < 26; i++) {
                maxFreq[i] = Math.max(maxFreq[i], freq[i]);
            }
        }

        List<String> result = new ArrayList<>();
        // Check each word in words1
        for (String word : words1) {
            int[] freq = new int[26];
            for (char c : word.toCharArray())
                freq[c - 'a']++;
            boolean isUniversal = true;
            for (int i = 0; i < 26; i++) {
                if (freq[i] < maxFreq[i]) {
                    isUniversal = false;
                    break;
                }
            }
            if (isUniversal)
                result.add(word);
        }

        return result;
    }
}

File: Algorithms/src/916. Word Subsets/Code/solution.cpp
class Solution
{
public:
    vector<string> wordSubsets(vector<string> &words1, vector<string> &words2)
    {
        vector<int> maxFreq(26, 0);

        // Precompute the maximum frequency for each character in words2
        for (const string &word : words2)
        {
            vector<int> freq(26, 0);
            for (char c : word)
                freq[c - 'a']++;
            for (int i = 0; i < 26; ++i)
            {
                maxFreq[i] = max(maxFreq[i], freq[i]);
            }
        }

        vector<string> result;
        // Check each word in words1
        for (const string &word : words1)
        {
            vector<int> freq(26, 0);
            for (char c : word)
                freq[c - 'a']++;
            bool isUniversal = true;
            for (int i = 0; i < 26; ++i)
            {
                if (freq[i] < maxFreq[i])
                {
                    isUniversal = false;
                    break;
                }
            }
            if (isUniversal)
                result.push_back(word);
        }

        return result;
    }
};
